### PR TITLE
fix: add check for empty response

### PR DESCRIPTION
**Pull request by saffaalvi**
Resolves https://github.com/StatCan/jupyter-apis/issues/33

- Checks to see if response from API is not empty
- Fix cost table message to be more descriptive (over the last 24 hours instead of per day) as it is confusing to some users

### CODE REVIEW

File: frontend/src/app/main-table/cost-table/cost-table.component.html [hunk 1]
```diff
@@ -5,23 +5,29 @@
...
 
   <p>
-    Estimated cost per day of notebook server resources – summed values may not
+    Estimated cost of notebook server resources over the last 24 hours – summed values may not
     match total due to rounding
   </p>
 
   <mat-divider></mat-divider>
-  <table style="width: auto">
+  <table *ngIf="aggregatedCost!==null && aggregatedCost.data!==undefined; else serverError" style="width: auto">
```
Comment (brendangadd) [inline frontend/src/app/main-table/cost-table/cost-table.component.html:13]: Looks like `aggregatedCost` is `null` until a successful response occurs. So this would result in a failure message to the user while waiting for a response (which may succeed).

File: frontend/src/app/main-table/cost-table/cost-table.component.html [hunk 2]
```diff
@@ -5,23 +5,29 @@
   </div>
 
   <p>
-    Estimated cost per day of notebook server resources – summed values may not
+    Estimated cost of notebook server resources over the last 24 hours – summed values may not
```
Comment (brendangadd) [inline frontend/src/app/main-table/cost-table/cost-table.component.html:8]: What's the reason for this change?

**Review by brendangadd — CHANGES_REQUESTED** [review]

File: frontend/src/app/main-table/cost-table/cost-table.component.html (hunk 2 above)
Comment (saffaalvi) [reply frontend/src/app/main-table/cost-table/cost-table.component.html:8]: @ca-scribner and I felt it was confusing to some users as it implied it was the cost of resources for the current day, so if a new resource was added, user might think the cost should update.

Comment (ca-scribner) [reply frontend/src/app/main-table/cost-table/cost-table.component.html:8]: yeah the estimate given here isn't a forward-looking per-day cost estimate (I don't think?), but instead a backward 24-hour assessment of what you used.  So if today I run nothing, then tomorrow I turn on a GPU and lots of expensive stuff and immediately look at kubecost, it'll still say "Estimated cost per day of notebook server resources = $0".  But tomorrow if I look again, it'll be $$$.  So the wording change was suggested to make it clear we're estimating based on history, not current running resources

Even better would be if we did our own forward-looking estimate based on a shorter timeframe if that's possible.  for example, do something like "last minute of costs * 24 * 60 * 60 = estimated cost per day based on your usage in the last minute"

...unless I misunderstood kubecost's output.  If it is forward looking already, then the old wording is great

Comment (brendangadd) [reply frontend/src/app/main-table/cost-table/cost-table.component.html:8]: :+1:  `Estimated -> Approximate`

File: frontend/src/app/services/kubecost.service.ts [hunk 3]
```diff
@@ -48,18 +49,24 @@ export class KubecostService {
...
   }
 
-  private handleError(
-    error: HttpErrorResponse | AggregateCostResponse
-  ): Observable<never> {
+  private handleError(error: HttpErrorResponse | AggregateCostResponse | Resp): Observable<never> {
+    if (this.isResp(error)) {
+      return throwError(new Error(error.log));
+    }
     return throwError(new Error(error.message));
```
Comment (brendangadd) [inline frontend/src/app/services/kubecost.service.ts]:
```suggestion
    const message = this.isResp(error) ? error.log : error.message;
    return throwError(new Error(message));
```

File: frontend/src/app/services/kubecost.service.ts [hunk 4]
```diff
@@ -48,18 +49,24 @@ export class KubecostService {
...
         tap(res => this.handleBackendError(res)),
         catchError(err => this.handleError(err))
-      );
+      ) as Observable<AggregateCostResponse>;
   }
 
-  private handleBackendError(response: AggregateCostResponse) {
-    if (response.code < 200 || response.code >= 300) {
+  private handleBackendError(response: AggregateCostResponse | Resp) {
+    if (this.isResp(response) || response.code < 200 || response.code >= 300) {
```
Comment (brendangadd) [inline frontend/src/app/services/kubecost.service.ts:56]: This is nice.

File: frontend/src/app/main-table/cost-table/cost-table.component.ts [hunk 5]
```diff
@@ -9,27 +15,22 @@ import { AggregateCostResponse } from 'src/app/services/kubecost.service';
...
-      if (Object.keys(this.aggregatedCost).includes('log')) {
-        this.message = "Failed to retrieve cost information."
-        return false;
-      }
-      else {
-        return true;
-      }
-    }
+  getStatus(): AsyncStatus {
+    if (this.currNamespace=='' || this.aggregatedCost==null) {
```
Comment (brendangadd) [inline frontend/src/app/main-table/cost-table/cost-table.component.ts:26]: I don't think this component should know about the current namespace. That is a concern of the parent component. If the parent wants to reset the state of this table when the namespace changes, then it can set the `aggregatedCost` parameter to `null`.

File: frontend/src/app/main-table/cost-table/cost-table.component.scss [hunk 6]
```diff
@@ -2,6 +2,15 @@ p {
   padding-left: 24px;
   padding-right: 16px;
 }
+.waiting {
+  padding: 20px;
+  padding-left: 24px;
+}
```
Comment (brendangadd) [inline frontend/src/app/main-table/cost-table/cost-table.component.scss:8]: Is this used?

**Review by brendangadd — CHANGES_REQUESTED** [review]
Your changes need formatting. @wg102 can show you how to make Prettier only format the parts you've changed when you save in VS Code, which will avoid adding ~20 files worth of formatting to this PR. :)

File: frontend/src/app/main-table/cost-table/cost-table.component.scss (hunk 6 above)
Comment (saffaalvi) [reply frontend/src/app/main-table/cost-table/cost-table.component.scss:8]: not anymore - removed!

**Review by brendangadd — DISMISSED** [review]
